### PR TITLE
Module3 was done

### PR DESCRIPTION
**Pull request by Gorylko**
All tasks except the first (it was sent earlier)

### CODE REVIEW

File: Module3/Task2/Program.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            }
+
+            var numberContainer = new List<int>();
+
+            for(int i = 0; i < amount; i++)
+            {
+                numberContainer.Add((i + 1) * 2);
+            }
+
+            return numberContainer;
```
Comment (NEDRUT) [inline Module3/Task2/Program.cs:41]: What about of using array instead List?

File: Module3/Task2/Program.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        private static void BugReport()
+        {
+            Console.WriteLine("User input is not valid");
+        }
```
Comment (NEDRUT) [inline Module3/Task2/Program.cs:47]: Let's reuse this method
Wrong method name. Looks like should be named RaiseError for example

File: Module3/Task3/Extensions/CollectionExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+            var stringBuilder = new StringBuilder();
+
+            foreach (var el in collection)
+            {
+                stringBuilder.Append(el.ToString() + " ");
+            }
+
+            return stringBuilder.ToString();
+        }
+    }
```
Comment (NEDRUT) [inline Module3/Task3/Extensions/CollectionExtensions.cs:24]: Reuse from task2

File: Module3/Task3/Program.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            }
+            else
+            {
+                BugReport();
+            }
+
+            Console.ReadKey();
+        }
+
+        public static IEnumerable<int> GetFibonacciNums()
```
Comment (NEDRUT) [inline Module3/Task3/Program.cs:29]: Put countOfNums in parameters of method and remove using Linq

File: Module3/Task3/Program.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            Console.WriteLine("Enter an integer greater than zero");
+            if (int.TryParse(Console.ReadLine(), out int number) &&
+                number > 0)
+            {
+                Console.Clear();
+                var result = GetFibonacciNums().TakeWhile((el, index) => index < number);
```
Comment (NEDRUT) [inline Module3/Task3/Program.cs:17]: Also, what about arrays?

File: Module3/Task2/Program.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+            if(int.TryParse(Console.ReadLine(), out int number) &&
+                number > 0)
+            {
+                Console.Clear();
+                Console.WriteLine(GetFirstEvenNaturalNumbers(number).AsString());
+            }
+            else
+            {
+                BugReport();
```
Comment (NEDRUT) [inline Module3/Task2/Program.cs:21]: Make possible to reinput data in case of errors

File: Module3/Task4/Parsers/DoubleParser.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+            if (
+                !double.TryParse(stringInput, NumberStyles.Any, CultureInfo.CurrentCulture, out number) &&
+                !double.TryParse(stringInput, NumberStyles.Any, CultureInfo.GetCultureInfo("en-US"), out number) &&
+                !double.TryParse(stringInput, NumberStyles.Any, CultureInfo.InvariantCulture, out number))
+            {
+                return false;
+            }
+            return true;
+        }
+    }
```
Comment (NEDRUT) [inline Module3/Task4/Parsers/DoubleParser.cs:40]: Remove

File: Module3/Task4/Program.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        static void Main(string[] args)
+        {
+            var reverser = new NumberReverser();
+
+            Console.WriteLine("Enter the number");
+
+            var userString = Console.ReadLine();
+
+            if (DoubleParser.TryParse(userString, out double parsedDouble))
```
Comment (NEDRUT) [inline Module3/Task4/Program.cs:16]: Can be used only double.TryParse(stringInput, NumberStyles.Any, CultureInfo.InvariantCulture, out number)

File: Module3/Task4/NumberReverser.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            if (number < 0)
+            {
+                _isNegative = true;
+                number *= -1;
+            }
+
+            number = this._isNegative? 
+                - double.Parse(new string(number.ToString().ToCharArray().Reverse().ToArray()))
+                : double.Parse(new string(number.ToString().ToCharArray().Reverse().ToArray()));
+
```
Comment (NEDRUT) [inline Module3/Task4/NumberReverser.cs:44]: Use cycle FOR to reverse.

File: Module3/Task4/NumberReverser.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
+﻿using System.Linq;
+
+namespace Task4
+{
+    public class NumberReverser
```
Comment (NEDRUT) [inline Module3/Task4/NumberReverser.cs:5]: 2 methods with the same structure. Difference only in parsed type.

Comment (NEDRUT) [reply Module3/Task4/NumberReverser.cs:5]: U can make method with next signature for example  public void Reverse(string inputValue)

File: Module3/Task6/Program.cs [hunk 11]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+
+namespace Task6
+{
+    class Program
+    {
+        private const int ArrayLength = 10;
```
Comment (NEDRUT) [inline Module3/Task6/Program.cs:7]: ARRAY_LENGTH

File: Module3/Task5/Initializers/UserInputInitializer.cs [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+using Task5.Initializers;
+using Task5.Initializers.Params;
+
+namespace Task5.Initializers
+{
+    public class UserInputInitializer
+    {
+        public UserInputResult<UserInputParams> InitializeData()
+        {
+
```
Comment (NEDRUT) [inline Module3/Task5/Initializers/UserInputInitializer.cs:15]: empty line

File: Module3/Task5/NumberRemover.cs [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+            var strNumber = number.ToString();
+
+            strNumber = strNumber.Replace(numberToDelete.ToString(), "");
+
+            if(strNumber == "")
+            {
+                return 0;
+            }
+
+            return int.Parse(strNumber);
```
Comment (NEDRUT) [inline Module3/Task5/NumberRemover.cs:16]: What about arrays and cycles?

File: Module3/Task7/Program.cs [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+                {
+                    resultNums.Add(array[i + 1]);
+                }
+            }
+
+            return resultNums;
+        }
+
+        private static void LogCollection<T>(IEnumerable<T> collection)
+        {
```
Comment (NEDRUT) [inline Module3/Task7/Program.cs:44]: If collection = null

File: Module3/Task8/EquationCulculator.cs [hunk 15]
```diff
@@ -0,0 +1,60 @@
+﻿using System;
+
+namespace Task8
+{
+    public class EquationCulculator
+    {
+        public EquationCulculator() { }
```
Comment (NEDRUT) [inline Module3/Task8/EquationCulculator.cs:7]: remove

File: Module3/Task8/Initializers/SpiralArrayInitializer.cs [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        private const int PrintInterval = 4;
+
+        private int _counter;
+        private Point? _currentPosition;
+        private Direction _currentDirection;
+
+        public SpiralArrayInitializer(int[,]array)
+        {
+            this.Array = array ?? throw new NullReferenceException(nameof(array));
```
Comment (NEDRUT) [inline Module3/Task8/Initializers/SpiralArrayInitializer.cs:16]: Bad idea to throw errors in constructor body

File: Module3/Task8/Initializers/Point.cs [hunk 17]
```diff
@@ -0,0 +1,15 @@
+﻿namespace Task8.Initializers
+{
+    public struct Point
```
Comment (NEDRUT) [inline Module3/Task8/Initializers/Point.cs:3]: Why struct?

File: Module3/Task8/Initializers/SpiralArrayInitializer.cs [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+                return false;
+            }
+            var position = this._currentPosition;
+            while(position != null)
+            {
+                this._counter++;
+                this.Array[position.Value.Row, position.Value.Column] = _counter;
+                PrintElement(this._counter, position);
+                this._currentPosition = position;
+                Thread.Sleep(100);
```
Comment (NEDRUT) [inline Module3/Task8/Initializers/SpiralArrayInitializer.cs:39]: Why?

**Review by NEDRUT — APPROVED** [review]

File: Module3/Task8/Initializers/SpiralArrayInitializer.cs (hunk 16 above)
Comment (Gorylko) [reply Module3/Task8/Initializers/SpiralArrayInitializer.cs:16]: Why?

File: Module3/Task8/Initializers/SpiralArrayInitializer.cs (hunk 18 above)
Comment (Gorylko) [reply Module3/Task8/Initializers/SpiralArrayInitializer.cs:39]: To demonstrate the application

File: Module3/Task8/Initializers/Point.cs (hunk 17 above)
Comment (Gorylko) [reply Module3/Task8/Initializers/Point.cs:3]: I want to be sure that the object will not be changed in the methods

**Pull request merged**